### PR TITLE
Fix/upgrade-caas-model

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

This PR fixes a few issues on CAAS models upgrading from 2.7 to 2.8:

-  remove legacy charm dir symlink for units deployed before 2.8;
- fix install hook failure caused by a Mongo assertion failed for the new field (`upgrade-counter`)in `podSpecs` collection;
- enhance the process of init container upgrading from 2.7 to 2.8;

## QA steps

1. bootstrap to 2.7 and deploy some workload;
2. upgrade controller then model to 2.8;
3. monitor the upgrading process;

## Documentation changes

None

## Bug reference

https://bugs.launchpad.net/juju/+bug/1874506
https://bugs.launchpad.net/juju/+bug/1877025

### CODE REVIEW

File: version/version.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ import (
 // The presence and format of this constant is very important.
 // The debian/rules build recipe uses this value for the version
 // number of the release package.
-const version = "2.8-rc1"
+const version = "2.8-rc16"
```
Comment (hpidcock) [inline version/version.go:24]: Make sure not to merge this

Comment (ycliuhw) [reply version/version.go:24]: yeah, will revert to beta1

File: apiserver/facades/agent/uniter/uniter.go [hunk 2]
```diff
@@ -3478,7 +3478,7 @@ func (u *UniterAPI) CommitHookChanges(args params.CommitHookChangesArgs) (params
 	return params.ErrorResults{Results: res}, nil
 }
 
-func (u *UniterAPI) commitHookChangesForOneUnit(unitTag names.UnitTag, changes params.CommitHookChangesArg, canAccessUnit, canAccessApp common.AuthFunc) error {
+func (u *UniterAPI) commitHookChangesForOneUnit(unitTag names.UnitTag, changes params.CommitHookChangesArg, canAccessUnit, canAccessApp common.AuthFunc) (err error) {
```
Comment (wallyworld) [inline apiserver/facades/agent/uniter/uniter.go:3481]: do we need this change?

File: cmd/jujud/agent/caasunitinit.go [hunk 3]
```diff
@@ -168,6 +168,7 @@ func (c *CAASUnitInitCommand) Run(ctx *cmd.Context) (errOut error) {
 	} else if err != nil {
 		return errors.Annotatef(err, "failed to stat %s", jujucPath)
 	}
+	logger.Criticalf("CAASUnitInitCommand os.Environ -> %q", os.Environ())
```
Comment (wallyworld) [inline cmd/jujud/agent/caasunitinit.go:171]: delete me

File: worker/uniter/runner/context/context.go [hunk 4]
```diff
@@ -1147,8 +1147,10 @@ func (ctx *HookContext) doFlush(process string) error {
 // we'll still trigger a change to a counter on the podspec so that we can
 // ensure any other charm changes (eg storage) are acted on.
 func (ctx *HookContext) addCommitHookChangesForCAAS(builder *uniter.CommitHookParamsBuilder, process string) error {
-	if ctx.podSpecYaml == nil && ctx.k8sRawSpecYaml == nil && process != string(hooks.UpgradeCharm) {
-		// No ops for any situation unless any k8s spec needs to be set or "upgrade-charm" was run.
+	if ctx.podSpecYaml == nil && ctx.k8sRawSpecYaml == nil {
```
Comment (wallyworld) [inline worker/uniter/runner/context/context.go:1150]: As discussed need to fix this check so that leaders still go through will nill spec

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[https://bugs.launchpad.net/juju/+bug/1877025](https://trello.com/c/vKZTjkFo/1117-https-bugslaunchpadnet-juju-bug-1877025)

**Pull request merged**